plugins/slick.cellselectionmodel.js

```diff
@@ -82,15 +82,20 @@
       return !areDifferent;
     }
 
-    function setSelectedRanges(ranges) {
+    function setSelectedRanges(ranges, caller) {
       // simple check for: empty selection didn't change, prevent firing onSelectedRangesChanged
       if ((!_ranges || _ranges.length === 0) && (!ranges || ranges.length === 0)) { return; }
 
       // if range has not changed, don't fire onSelectedRangesChanged
       var rangeHasChanged = !rangesAreEqual(_ranges, ranges);
 
       _ranges = removeInvalidRanges(ranges);
-      if (rangeHasChanged) { _self.onSelectedRangesChanged.notify(_ranges); }
+      if (rangeHasChanged) { 
+        // provide extra "caller" argument through SlickEventData to avoid breaking pubsub event that only accepts an array of selected range
+        var eventData = new Slick.EventData();
+        Object.defineProperty(eventData, 'detail', { writable: true, configurable: true, value: { caller: caller || "SlickCellSelectionModel.setSelectedRanges" } });
+        _self.onSelectedRangesChanged.notify(_ranges, eventData); 
```

Comment:
@6pac could you review and provide feedback to this PR please. 
I had to use cheating code to get this working, by passing extra data to an SlickEventData (which was not made for that), but I don't see other ways because I did not want to introduce breaking change. The issue I had and why I resolve to cheat is because `onSelectedRowsChanged` and `onSelectedRangesChanged` only accept an array of SlickRange or row number, so it's different compare to regular event that we typically return an object as `args` while these are straight array (I can't extend them to provide extra properties like I wanted). Are you ok with the change? Do you have any other suggestions?

Reply:
From what I can see, the calling pattern for ```onSelectedRowsChanged``` in the core grid is:

      if (simpleArrayEquals(previousSelectedRows, selectedRows)) {
        trigger(
            self.onSelectedRowsChanged, 
            {   rows: getSelectedRows(), 
                previousSelectedRows: previousSelectedRows
            }, 
            e);
      }

So this uses the 'args-is-an-object' convention, and you should be able to just add the property.
For ```onSelectedRangesChanged```, it appears only to be used in cellselectionmodel and rowselectionmodel (though of course it may also be called by user code), with the call:

      _self.onSelectedRangesChanged.notify(_ranges);

This doesn't follow the convention of 'args-is-an-object', which really should be corrected, but that would be a breaking change. The only non-breaking way I can think of is to take advantage of the fact that in js an array is also an object. So we could do, for example:

      var args = _ranges;
      args.ranges = _ranges;
      args.caller = caller;
      _self.onSelectedRangesChanged.notify(args);

a bit of a hack, but technically correct. It may be confusing when looking in the debugger, but I think a comment at the call would be enough to demystify that.

Reply:
hmm I can give a try next week but I'm not crazy about this approach and I don't think that would work because the other side is really expecting an array not an object, but since nearly everything is an object in JS we never know, I'll give a try next week. If that doesn't work, my approach, even if not ideal, is still working and for the most part this code is used only internally (to communicate between SlickGrid & the plugins) so we might be ok since it's not that expose to the outside user.

Reply:
Sure. What I'm suggesting is that in internal code we treat it as an object. Setting it as an array is only to maintain backwards compatibilty for legacy user calls.

Reply:
Given that my proposed solution is a hack, I think yours works equally well, so either-or. The issue with ```onSelectedRangesChanged``` does need to go on the list for when we finally release a breaking version.

Reply:
@6pac yeah I would rather merge it as it is, I know it works with use case in our Salesforce environment and so I would rather use that for now. It's mostly an internal thing but you never know when someone hooks into any of the `onSelectedRangesChanged` events (I'm sure there's someone in the world doing that) so I would rather not break them and my solution even though hacky as well isn't too bad and it works without breaking anything

Reply:
OK, so we are ready to merge this and release?

Reply:
wait give me another 5min, I'm giving a try with your hack which is basically this [SO](https://stackoverflow.com/questions/9952126/add-a-property-to-a-javascript-array)

Reply:
it seems to work too, but I'm not sure how that would work with TypeScript. I would say let's merge with the code I have now (it works with TypeScript in my lib) and we can keep it as a 2nd option. Let me merge my PR then

Reply:
Yeah, I don't think Typescript would like it. As they say in the SO post, it's not that it won't work, it's just quite an unusual pattern and might be quite confusing for some developers. OK, willl pull and release.